### PR TITLE
using tenacity in download_binary function

### LINKED ISSUE #392: Use tenacity in download_binary
https://github.com/mozilla/code-coverage/blob/e621d18c259773cafc466b092396ede9ff69b566/report/firefox_code_coverage/codecoverage.py#L70

### PR DESCRIPTION
**Pull request by NitishNaineni**
Have used the tenacity library to implement the retry functionality into the download_binary function

Fixes #392.

### CODE REVIEW

File: report/firefox_code_coverage/codecoverage.py [hunk 1]
```diff
@@ -69,27 +70,28 @@ def _save_tasks(response):
 
 def download_binary(url, path, retries=5):
     """Download a binary file from an url"""
-    for i in range(1, retries + 1):
+    @tenacity.retry(stop=tenacity.stop_after_attempt(retries),
```
Comment (marco-c) [inline report/firefox_code_coverage/codecoverage.py]: Set this annotation directly on `download_binary`, so you don't have to define an inner function.

File: report/firefox_code_coverage/codecoverage.py [hunk 2]
```diff
@@ -69,27 +70,28 @@ def _save_tasks(response):
...
                 os.remove(path)
             except OSError:
                 pass
-
-            if i == retries:
-                raise Exception(
-                    "Download failed after {} retries - {}".format(retries, url)
+            
+            raise Exception(
+                    "Download failed after {} retries - {}".format(retry_function.retry.statistics["attempt_number"], url)
```
Comment (marco-c) [inline report/firefox_code_coverage/codecoverage.py]: We can remove the number of retries from the exception message to simplify this code.

File: report/firefox_code_coverage/codecoverage.py [hunk 3]
```diff
@@ -69,27 +70,28 @@ def _save_tasks(response):
...
         try:
             artifact = requests.get(url, stream=True)
             artifact.raise_for_status()
 
             with open(path, "wb") as f:
                 for chunk in artifact.iter_content(chunk_size=8192):
                     f.write(chunk)
-            break
-        except:  # noqa: E722
+        except:
```
Comment (marco-c) [inline report/firefox_code_coverage/codecoverage.py]: Use `except Exception` to avoid linting issues

File: report/firefox_code_coverage/codecoverage.py (hunk 2 above)
Comment (marco-c) [reply report/firefox_code_coverage/codecoverage.py]: Even more simply, just do `raise` to re-raise the previous exception.

File: report/firefox_code_coverage/codecoverage.py (hunk 1 above)
Comment (NitishNaineni) [reply report/firefox_code_coverage/codecoverage.py]: But then how do we input the retries number to the decorator? I have written an inner function in order to pass the retries to the decorator. I am sorry if this is a silly question, this is my first time using decorators.

Comment (marco-c) [reply report/firefox_code_coverage/codecoverage.py]: It's not a silly question at all! We can simplify the function and just remove the argument and hardcode the number of retries.

**Comment by marco-c** [comment]
Why did you close the PR?

**Comment by NitishNaineni** [comment]
@marco-c  sorry i am figuring out how to use pull requests

File: report/firefox_code_coverage/codecoverage.py (hunk 1 above)
Comment (NitishNaineni) [reply report/firefox_code_coverage/codecoverage.py]: sure i have made the changes

**Comment by marco-c** [comment]
> @marco-c sorry i am figuring out how to use pull requests

Oh, ok! No worries ;)

File: report/firefox_code_coverage/codecoverage.py (hunk 2 above)
Comment (NitishNaineni) [reply report/firefox_code_coverage/codecoverage.py]: i have made this change

**Comment by marco-c** [comment]
I suggest using pre-commit to handle formatting for you

**Comment by marco-c** [comment]
You'll need to add `tenacity` to the requirements (https://github.com/mozilla/code-coverage/blob/master/report/requirements.txt)

**Comment by NitishNaineni** [comment]
i have used per-commit and also added `tenacity` to requirements.txt

**Comment by NitishNaineni** [comment]
@marco-c  can you tell me why its failing the tests? I will try to fix it

File: bot/tests/test_notifier.py [hunk 4]
```diff
@@ -50,5 +50,5 @@ def test_notification(mock_secrets, mock_taskcluster, mock_phabricator, fake_hg_
     mail = notify_email(revision, stack, changesets_coverage)
     assert (
         mail
-        == "* [Commit [(b'M', b'file')]Differential Revision: https://phabricator.services.mozilla.com/D2](https://phabricator.services.mozilla.com/D2): 0 covered out of 6 added.\n"  # noqa
+        == "* [Commit [(b'M', b'file')]Differential Revision: https://phabricator.services.mozilla.com/D2](https://phabricator.services.mozilla.com/D2): 0 covered out of 6 added.\n"
```
Comment (marco-c) [inline bot/tests/test_notifier.py]: Please revert this change and the change in bot/tests/test_taskcluster.py, as they are unrelated.

**Comment by marco-c** [comment]
> @marco-c can you tell me why its failing the tests? I will try to fix it

It seems the tests are not failing, the linting is failing (you can look at the logs on Taskcluster by clicking on "Details").

**Comment by NitishNaineni** [comment]
@marco-c  
i have gone through the logs this seems to be the issue
```
Fix requirements.txt.....................................................Failed
- hook id: requirements-txt-fixer
- exit code: 1
- files were modified by this hook
```
i don't understand the issue here. Can you tell me whats wrong?

**Comment by marco-c** [comment]
Probably the order of the requirements is wrong.
If you use pre-commit, all these kinds of errors will be automatically fixed for you, and you wouldn't have to care about them.

**Comment by marco-c** [comment]
@NitishNaineni you removed the branch, which closed the PR. Could you restore it?

**Comment by NitishNaineni** [comment]
@marco-c 
i am not able to do any git operations on it due to unstaged files. It seems i broke something so i am going to recreate and put another pull request. Sorry for the inconvenience

**Comment by marco-c** [comment]
If you create a new branch with the same name and force push it, you can update the current pull request without opening a new one.

**Comment by NitishNaineni** [comment]
sorry not able to reopen it
